### PR TITLE
feat: upgrade to eslint v9 and fix breaking changes

### PR DESCRIPTION
**Pull request by erik-slovak**
BREAKING CHANGE: remove matching patterns from configs and export individual rule sets

### CODE REVIEW

File: eslint.config.js [hunk 1]
```diff
@@ -1,16 +1,50 @@
-const txoConfig = require('./lib/index.js')
+const tseslint = require('typescript-eslint')
```
Comment (rostislav-simonik) [inline eslint.config.js:1]: can we preserve full name of package typescriptEslint

File: eslint.config.js [hunk 2]
```diff
@@ -1,16 +1,50 @@
...
+const stylisticPlugin = require('@stylistic/eslint-plugin')
```
Comment (rostislav-simonik) [inline eslint.config.js:2]: eslintPluginStylistic

File: eslint.config.js [hunk 3]
```diff
@@ -1,16 +1,50 @@
...
+
+const {
+  stylisticConfig,
+  ignoreList,
+  typescriptConfigList,
+} = require('./lib/index.js')
```
Comment (rostislav-simonik) [inline eslint.config.js:8]:
```suggestion
} = require('./lib')
```

File: eslint.config.js [hunk 4]
```diff
@@ -1,16 +1,50 @@
...
 
 /** @type {import('eslint').Linter.FlatConfig[]} */
-const config = [
-  ...txoConfig.default,
+const config = tseslint.config(
+  // TODO: remove after migrating to prettier
+  {
+    files: ['**/*.ts'],
+    extends: [
+      {
```
Comment (rostislav-simonik) [inline eslint.config.js:16]: do we need this

File: eslint.config.js [hunk 5]
```diff
@@ -1,16 +1,50 @@
...
+        languageOptions: {
+          parserOptions: {
+            projectService: true,
+          },
+        },
+      },
+      {
```
Comment (rostislav-simonik) [inline eslint.config.js:23]: do we need this

File: eslint.config.js [hunk 6]
```diff
@@ -1,16 +1,50 @@
...
+        plugins: {
+          '@stylistic': stylisticPlugin,
+        }
+      },
+      ...typescriptConfigList,
+      // TODO: remove after migrating to prettier
+      stylisticConfig,
+      {
+        rules: {
+          '@typescript-eslint/no-magic-numbers': 'off',
```
Comment (rostislav-simonik) [inline eslint.config.js:33]: do we need to really disable magic numbers

File: src/configs/typescript.ts [hunk 7]
```diff
@@ -60,10 +60,9 @@ export const typescriptRules: Linter.RulesRecord = {
     },
   }],
   '@typescript-eslint/consistent-type-definitions': 'off',
-  '@typescript-eslint/ban-types': [
+  '@typescript-eslint/no-restricted-types': [
```
Comment (rostislav-simonik) [inline src/configs/typescript.ts:63]: let's check what is in eslint-config-love

File: src/index.ts [hunk 8]
```diff
@@ -1,65 +1,85 @@
...
+  configs as typescriptEslintConfigs,
+  parser,
+} from 'typescript-eslint'
+import type { TSESLint } from '@typescript-eslint/utils'
+import eslintComments from '@eslint-community/eslint-plugin-eslint-comments/configs'
+import { flatConfigs as eslintImportFlatConfigs } from 'eslint-plugin-import'
+import eslintImportRecommendedFlatConfig from 'eslint-plugin-import/config/flat/recommended'
+import stylistic from '@stylistic/eslint-plugin'
+import jestPlugin from 'eslint-plugin-jest'
+import loveConfig from 'eslint-config-love'
```
Comment (rostislav-simonik) [inline src/index.ts:18]: let's unify naming convention to follow the import name if it's scoped and it's eslint-plugin or eslint-config let's add scope as suffix

File: src/index.ts [hunk 9]
```diff
@@ -1,65 +1,85 @@
...
+export const typescriptConfigList: TSESLint.FlatConfig.ConfigArray = [
+  linterOptionsConfig,
+  ...typescriptEslintConfigs.recommendedTypeChecked,
+  ...typescriptEslintConfigs.stylisticTypeChecked,
+  loveConfig,
+  eslintComments.recommended,
+  {
+    // manually create the `eslint-plugin-import` recommended config because `eslint-config-love` already defines it in `plugins.import` which causes a conflict
+    name: 'import/recommended',
+    ...eslintImportRecommendedFlatConfig,
```
Comment (rostislav-simonik) [inline src/index.ts:49]: Check if it's already in eslint-love-recommended

File: src/index.ts [hunk 10]
```diff
@@ -1,65 +1,85 @@
...
+      strict: 'error',
+      ...typescriptRules,
+      ...loveRules,
+      ...importRules,
+      ...eslintCommentsRules,
+    },
+    settings: {
+      'import/parsers': {
+        espree: ['.js', '.cjs', '.mjs', '.jsx'],
+      },
```
Comment (rostislav-simonik) [inline src/index.ts:69]: do we need this ?

File: src/types/eslint-plugin-import/config/flat/recommended.d.ts [hunk 11]
```diff
@@ -0,0 +1,4 @@
+import type { Linter } from 'eslint'
```
Comment (rostislav-simonik) [inline src/types/eslint-plugin-import/config/flat/recommended.d.ts:1]: let's move into index.d.ts and use declare mode with subpath

File: test/Sample.ts [hunk 12]
```diff
@@ -1,104 +0,0 @@
-/**
```
Comment (rostislav-simonik) [inline test/Sample.ts:1]: let's leave one rule per plugin to have a smoke test.

File: src/types/@eslint-community/eslint-plugin-eslint-comments/configs.d.ts [hunk 13]
```diff
@@ -0,0 +1,12 @@
+// TODO: delete this after https://github.com/eslint-community/eslint-plugin-eslint-comments/pull/246 is merged
```
[the remark this answers is not in the document]
Comment (erik-slovak) [reply src/types/@eslint-community/eslint-plugin-eslint-comments/configs.d.ts:1]: I'm getting errors when trying to export from index.d.ts: `Exports and export assignments are not permitted in module augmentations.`

File: src/index.ts [hunk 14]
```diff
@@ -1,65 +1,73 @@
...
 **/
 
-import type { Linter } from 'eslint'
 import globals from 'globals'
+import {
+  configs as typescriptEslintConfigs,
+  parser,
+} from 'typescript-eslint'
+import type { TSESLint } from '@typescript-eslint/utils'
+import eslintPluginEslintCommunityEslintComments from '@eslint-community/eslint-plugin-eslint-comments/configs'
```
Comment (rostislav-simonik) [inline src/index.ts:13]:
```suggestion
import eslintPluginEslintComments from '@eslint-community/eslint-plugin-eslint-comments/configs'
```

File: src/index.ts [hunk 15]
```diff
@@ -1,65 +1,73 @@
...
+import { flatConfigs as eslintPluginImportFlatConfigs } from 'eslint-plugin-import'
```
Comment (rostislav-simonik) [inline src/index.ts:14]:
```suggestion
import { flatConfigs as eslintPluginImportConfigList } from 'eslint-plugin-import'
```

**Review by rostislav-simonik — APPROVED** [review]

**Pull request merged**